### DISCUSSION TITLE
nonconst `Evaluate()` in `GradientProblem` / `FirstOrderFunction`

### DISCUSSION THREAD

**Discussion by cpp977** [ceres-solver/ceres-solver#818] [Q&A] [answered]
Hey ceres-team,

I found a use case where I need a nonconst version of `Evaluate()` in `GradientProblem` and `FirstOrderFunction` because:

1. I need to apply different approaches to compute the cost and the gradient depending on the measure of convergence and 
2. I need to reuse results from previous cost evauations. For that, I store some intermediate results in the `FirstOrderFunction` class but    during `Evaluate` I cannot update them for upcoming calls to `Evaluate()`. 

Maybe there is a workflow to achieve this within the current framework (ideally without global variables). Otherwise, is it conceivable to drop the `const` qualification of the `Evaluate()` function?

Thanks in advance for help!

**Comment by alexsmac** [comment, marked as answer]
We wouldn't change the API to support this, as the current use of `const` reflects our expectation of the contract to `FirstOrderFunction`, namely that it is idempotent.  Mathematically, this is what is required.  If you want to violate this condition, whilst it might work you should understand that what you are doing violates the underlying assumptions of the solver and so YMMV, in general though you should assume that this would be problematic.

If you wanted to do what you're describing anyway, you could do it via a pointer indirection to implementation (i.e. PIMPL), something like the following:
```
struct Impl {
  MutableEvaluate(...);
};

class Cost final : public FirstOrderFunction {
  public:
    virtual bool Evaluate(const double* const parameters,
                        double* cost,
                        double* gradient) const override {
       impl_->MutableEvaluate(...);
     }
  private:
     std::unqiue_ptr<Impl> impl_;
};
```
This will work, as the `impl_` `unique_ptr` itself is not being mutated, therefore the `const` nature of `Evaluate()` is preserved from the perspective of `Cost`, however it is being used to invoke a mutable operation on the implementation.
